packages/flutter/lib/src/material/data_table.dart

```diff
@@ -1080,7 +1099,7 @@ class DataTable extends StatelessWidget {
         sorted: dataColumnIndex == sortColumnIndex,
         ascending: sortAscending,
         overlayColor: effectiveHeadingRowColor,
-        semanticsHeadingLabel: column.heading != null ? column.heading! : '',
+        semanticsHeadingLabel: column.heading != null ? column.heading! : '${dataColumnIndex + 1}',
```

Comment:
why +1?

Reply:
Added comments to explain why. Basically for example for columns[0], when the heading is present we want it to be announced as 'Sorted by column $columnHeading' in screen readers announcements, if the heading is null we want  something like 'Sorted by column 1'.